### PR TITLE
[SPARK-36352][SQL] Spark should check result plan's output schema name

### PR DESCRIPTION
**Pull request by AngersZhuuuu**
### What changes were proposed in this pull request?
Spark should check result plan's output schema name

### Why are the changes needed?
In current code, some optimizer rule may change plan's output schema, since in the code we always use semantic equal to check output, but it may change the plan's output schema.
For example, for SchemaPruning, if we have a plan
```
Project[a, B]
|--Scan[A, b, c]
```
the origin output schema is `a, B`, after SchemaPruning. it become
```
Project[A, b]
|--Scan[A, b]
```
It change the plan's schema. when we use CTAS, the schema is same as query plan's output.
Then since we change the schema, it not consistent with origin SQL. So we need to check final result plan's schema with origin plan's schema

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
existed UT

### CODE REVIEW

**Comment by AngersZhuuuu** [comment]
ping @cloud-fan

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala [hunk 1]
```diff
@@ -174,8 +174,8 @@ class Analyzer(override val catalogManager: CatalogManager)
 
   private val v1SessionCatalog: SessionCatalog = catalogManager.v1SessionCatalog
 
-  override protected def isPlanIntegral(plan: LogicalPlan): Boolean = {
-    !Utils.isTesting || LogicalPlanIntegrity.checkIfExprIdsAreGloballyUnique(plan)
+  override protected def isPlanIntegral(input: LogicalPlan, result: LogicalPlan): Boolean = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala:177]: nit: I'd name them `previousPlan` and `currentPlan`

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 2]
```diff
@@ -46,10 +46,13 @@ abstract class Optimizer(catalogManager: CatalogManager)
...
   // - has globally-unique attribute IDs
-  override protected def isPlanIntegral(plan: LogicalPlan): Boolean = {
-    !Utils.isTesting || (plan.resolved &&
-      plan.find(PlanHelper.specialExpressionsInUnsupportedOperator(_).nonEmpty).isEmpty &&
-      LogicalPlanIntegrity.checkIfExprIdsAreGloballyUnique(plan))
+  override protected def isPlanIntegral(input: LogicalPlan, result: LogicalPlan): Boolean = {
+    !Utils.isTesting || (result.resolved &&
+      result.find(PlanHelper.specialExpressionsInUnsupportedOperator(_).nonEmpty).isEmpty &&
+      LogicalPlanIntegrity.checkIfExprIdsAreGloballyUnique(result) &&
+      input.output.zip(result.output).forall { case (in, out) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:53]: how abut `input.schema == result.schema`?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SchemaPruning.scala [hunk 3]
```diff
@@ -80,10 +80,10 @@ object SchemaPruning extends Rule[LogicalPlan] {
       // in dataSchema.
       if (countLeaves(dataSchema) > countLeaves(prunedDataSchema)) {
         val prunedRelation = leafNodeBuilder(prunedDataSchema)
-        val projectionOverSchema = ProjectionOverSchema(prunedDataSchema)
+        val projectionOverSchema = ProjectionOverSchema(prunedDataSchema,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SchemaPruning.scala:83]: cc @viirya

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 4]
```diff
@@ -4202,6 +4202,19 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
+
+  test("SPARK-36352: Optimizer should keep schema name not changed") {
+    withTable("t") {
+      sql("CREATE TABLE t1(a INT, b INT) USING PARQUET")
+      sql(
+        """
+          |INSERT INTO t1 VALUES
+          |(1, 2),
+          |(2, 3)""".stripMargin)
+      val df = sql("SELECT C FROM (SELECT a + b AS c FROM t1)")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4214]: This was broken before due to the issue in `ProjectionOverSchema`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 2 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: > how abut `input.schema == result.schema`?

Should use `schema.sameType()`?

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 4 above)
Comment (AngersZhuuuu) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala]: Removed

**Comment by AngersZhuuuu** [comment]
retest this please

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SchemaPruning.scala [hunk 5]
```diff
@@ -143,15 +145,15 @@ object SchemaPruning extends Rule[LogicalPlan] {
...
+    val newProjects = normalizedProjects.map(_.transformDown {
       case projectionOverSchema(expr) => expr
     }).map { case expr: NamedExpression => expr }
 
     if (log.isDebugEnabled) {
       logDebug(s"New projects:\n${newProjects.map(_.treeString).mkString("\n")}")
     }
 
-    Project(newProjects, projectionChild)
+    Project(restoreOriginalOutputNames(newProjects, projects.map(_.name)), projectionChild)
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/SchemaPruning.scala:156]: Seems okay.

**Comment by viirya** [comment]
> Why are the changes needed?
> Fix bug

Could you add something more to the description? Please describe it to let reviewers know what the bug is. Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 6]
```diff
@@ -46,10 +47,13 @@ abstract class Optimizer(catalogManager: CatalogManager)
   // - is still resolved
   // - only host special expressions in supported operators
   // - has globally-unique attribute IDs
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:49]: Could you also update the comment?

**Comment by viirya** [comment]
> How was this patch tested?
> existed UT

Could you add new test? For a bug fix PR, except for the rare cases that it is hard to do unit test, I think we usually need to provide a new test to show what the bug is and verify it is fixed after the PR.

**Comment by AngersZhuuuu** [comment]
> > How was this patch tested?
> > existed UT
> 
> Could you add new test? For a bug fix PR, except for the rare cases that it is hard to do unit test, I think we usually need to provide a new test to show what the bug is and verify it is fixed after the PR.

How about current?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala (hunk 6 above)
Comment (AngersZhuuuu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala:49]: Done

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master/3.2!

**Comment by cloud-fan** [comment]
@AngersZhuuuu can you help to open backport PRs for 3.1/3.0? thanks!

**Pull request closed without merging**